confgenerator/logging_processors.go

```diff
@@ -39,77 +44,182 @@ func init() {
 }
 
-// A LoggingProcessorParseRegex applies a regex to the specified field, storing the named capture groups as keys in the log record.
-type LoggingProcessorParseRegex struct {
-	ConfigComponent             `yaml:",inline"`
-	LoggingProcessorParseShared `yaml:",inline"`
+// A LoggingProcessorParseRegexSimple applies a regex to the specified field, storing the named capture groups as keys in the log record.
+// This was maintained in addition to the parse_regex_complex to ensure backward compatibility with any existing configurations
+type LoggingProcessorParseRegexSimple struct {
+	ConfigComponent        `yaml:",inline"`
+	fluentbit.ParserShared `yaml:",inline"`
+	Field                  string `yaml:"field,omitempty"`
 
 	Regex string `yaml:"regex,omitempty" validate:"required"`
 }
 
-func (r LoggingProcessorParseRegex) Type() string {
+func (r LoggingProcessorParseRegexSimple) Type() string {
 	return "parse_regex"
 }
 
-func (p LoggingProcessorParseRegex) Components(tag, uid string) []fluentbit.Component {
-	filter, parser := p.LoggingProcessorParseShared.Components(tag, uid)
+func (p LoggingProcessorParseRegexSimple) Components(tag, uid string) []fluentbit.Component {
+	parser, parserName := p.ParserShared.Component(tag, uid)
 	parser.Config["Format"] = "regex"
 	parser.Config["Regex"] = p.Regex
+
 	return []fluentbit.Component{
-		filter,
 		parser,
+		fluentbit.ParserFilterComponent(tag, p.Field, []string{parserName}),
+	}
+}
+
+type RegexParser struct {
+	Regex  string                 `yaml:"regex,omitempty" validate:"required"`
+	Parser fluentbit.ParserShared `yaml:",inline"`
+}
+
+// A LoggingProcessorParseRegex applies a set of regexes to the specified field, storing the named capture groups as keys in the log record.
+type LoggingProcessorParseRegex struct {
+	ConfigComponent `yaml:",inline"`
+	Field           string `yaml:"field,omitempty"`
+
+	Parsers []RegexParser `yaml:"parsers,omitempty" validate:"required"`
+}
+
+func (r LoggingProcessorParseRegex) Type() string {
+	return "parse_regex_complex"
+}
+
+func (p LoggingProcessorParseRegex) Components(tag, uid string) []fluentbit.Component {
+	components := []fluentbit.Component{}
+	parserNames := []string{}
+
+	for idx, parserConfig := range p.Parsers {
+		parser, parserName := parserConfig.Parser.Component(tag, fmt.Sprintf("%s.%d", uid, idx))
+		parser.Config["Format"] = "regex"
+		parser.Config["Regex"] = parserConfig.Regex
+		components = append(components, parser)
+		parserNames = append(parserNames, parserName)
+	}
+
+	components = append(components, fluentbit.ParserFilterComponent(tag, p.Field, parserNames))
+	return components
+}
+
+type MultilineRule struct {
+	StateName string `yaml:"state_name" validate:"required"`
+	Regex     string `yaml:"regex,omitempty" validate:"required"`
+	NextState string `yaml:"next_state" validate:"required"`
+}
+
+func (r MultilineRule) AsString() string {
+	return fmt.Sprintf(`"%s"    "%s"    "%s"`, r.StateName, r.Regex, r.NextState)
```

Comment:
What kind of escaping do these strings need (especially the regex)?

Reply:
The only thing I'm aware of is that they shouldn't have unescaped double quotes, otherwise they are just quoted regex literals (https://docs.fluentbit.io/manual/administration/configuring-fluent-bit/multiline-parsing)

Reply:
I think it's likely that some of these patterns will have double quotes in them, so does that mean we should add a call to ```strings.Replace(r.Regex, `"`, `\"`)``` here (is that the proper escaping syntax?)?

Reply:
I think that's good enough for now. Some concern that a developer could try to escape it themselves if they are familiar with the format, but should be caught pretty quickly in testing. Could write a more complex replace that would only escape if not already escaped but I think it makes sense to keep it simple for now

Reply:
I tested the `\"` escape on a deployment of the ops-agent I had set up for cassandra testing and it worked correctly.

Reply:
Let's make sure we have a test case in the actual application PR to cover this .